File: splitsy-backend/src/domain/users/value_objects/email.py
import re
from dataclasses import dataclass

from src.domain.common.errors import ValidationError
from src.domain.common.value_objects import ValueObject

EMAIL_REGEX = re.compile(r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$")


@dataclass(frozen=True)
class EmailProps:
    value: str


class Email(ValueObject[EmailProps]):
    def __init__(self, props: EmailProps) -> None:
        super().__init__(props)
        self._value = props.value

    @property
    def value(self) -> str:
        return self._value

    @staticmethod
    def create(value: str) -> "Email":
        if not isinstance(value, str):
            raise ValidationError(f"Invalid format: {value}")

        if not EMAIL_REGEX.match(value):
            raise ValidationError("Invalid email format")

        email = value.lower()

        return Email(EmailProps(value=email))

File: splitsy-frontend/src/domain/shared/value-objects/Email.ts
import { ValueObject } from "./ValueObject";

export interface EmailAttributes {
  value: string;
}

export class Email extends ValueObject<EmailAttributes> {
  private constructor(props: EmailAttributes) {
    super(props);
  }

  public static create(email: string): Email {
    if (this.isEmpty(email)) {
      throw new Error('Email is a required field');
    }

    if (!this.isValid(email)) {
      throw new Error('Invalid email address');
    }

    return new Email({ value: email });
  }

  static isValid(email: string): boolean {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
  }

  static isEmpty(email: string): boolean {
    return email.length === 0;
  }

  public get value(): string {
    return this.props.value;
  }
}
